Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/saf/file/AbstractFile.kt

```diff
@@ -0,0 +1,236 @@
+package com.github.adamantcheese.chan.core.saf.file
+
+import com.github.adamantcheese.chan.core.extension
+import com.github.adamantcheese.chan.core.saf.annotation.ImmutableMethod
+import com.github.adamantcheese.chan.core.saf.annotation.MutableMethod
+import java.io.File
+import java.io.InputStream
+import java.io.OutputStream
+
+/**
+ * An abstraction class over both the Java File and the new Storage Access Framework DocumentFile.
+ *
+ * Some methods are marked with [MutableMethod] annotation. This means that such method are gonna
+ * mutate the inner data of the [AbstractFile] (such as root or segments). Sometimes this behavior is
+ * not desirable. For example, when you have an AbstractFile representing some directory that may
+ * not even exists on the disk and you want to check whether it exists and if it does check some
+ * additional files inside that directory. In such case you may want to preserve the [AbstractFile]
+ * that represents that directory in it's original state. To do this you have to call the [clone]
+ * method on the file that represents the directory. It will create a copy of the file that you can
+ * safely work without worry that the original file may change.
+ *
+ * Other methods are marked with [ImmutableMethod] annotation. This means that those files create a
+ * copy of the [AbstractFile] internally and are safe to use without calling [clone]
+ * */
+abstract class AbstractFile(
```

Comment:
TODO: make this class thread-safe since we will need to move it to a background thread in the future.

Reply:
On the other hand, I will have to mark almost all of the methods as suspend and it's gonna be pretty hard to use them out of java. What needs to be done before this is The Great Java To Kotlin Conversion.